x/superfluid/keeper/migrate.go

```diff
@@ -1,99 +1,302 @@
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, err
+	}
+
+	// Force unlock, validate the provided sharesToMigrate, and exit the balancer pool.
+	// This will return the coins that will be used to create the concentrated liquidity position.
+	exitCoins, err := k.validateSharesToMigrateUnlockAndExitBalancerPool(ctx, sender, poolIdLeaving, preMigrationLock, gammSharesInLock, sharesToMigrate)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, err
+	}
+
+	// Create a full range (min to max tick) concentrated liquidity position, lock it, and superfluid delegate it.
+	positionId, amount0, amount1, liquidity, joinTime, concentratedLockId, err = k.clk.CreateFullRangePositionLocked(ctx, concentratedPool, sender, exitCoins, remainingLockTime)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, err
+	}
+	err = k.SuperfluidDelegate(ctx, sender.String(), concentratedLockId, intermediateAccount.ValAddr)
+	if err != nil {
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, err
+	}
+
+	// If there are remaining gamm shares, we must re-lock them.
+	remainingGammShares := gammSharesInLock.Sub(sharesToMigrate)
+	if !remainingGammShares.IsZero() {
+		// Create a new lock with the remaining gamm shares for the remaining lock time.
+		newLock, err := k.lk.CreateLock(ctx, sender, sdk.NewCoins(remainingGammShares), remainingLockTime)
+		gammLockId = newLock.ID
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, err
+		}
+
+		// If the gamm lock was previously superfluid bonded, superfluid delegate the gamm like normal
+		err = k.SuperfluidDelegate(ctx, sender.String(), gammLockId, valAddr)
+		if err != nil {
+			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, err
+		}
+	}
+
+	return positionId, amount0, amount1, liquidity, joinTime, gammLockId, concentratedLockId, nil
+}
+
+// migrateSuperfluidUnbondingBalancerToConcentrated migrates a user's superfluid unbonding balancer position to a superfluid unbonding concentrated liquidity position.
+// The function force unlocks and exits the balancer pool, creates a full range concentrated liquidity position, and locks it. If there are any remaining gamm shares, they are re-locked and begin unlocking where they left off.
+// The function returns the concentrated liquidity position ID, amounts of tokens in the position, the liquidity amount, join time, and IDs of the involved pools and locks.
+func (k Keeper) migrateSuperfluidUnbondingBalancerToConcentrated(ctx sdk.Context, sender sdk.AccAddress, poolIdLeaving, poolIdEntering uint64, preMigrationLock *lockuptypes.PeriodLock, gammSharesInLock sdk.Coin, lockId uint64, sharesToMigrate sdk.Coin, synthDenomBeforeMigration string, concentratedPool cltypes.ConcentratedPoolExtension, remainingLockTime time.Duration) (positionId uint64, amount0, amount1 sdk.Int, liquidity sdk.Dec, joinTime time.Time, gammLockId, concentratedLockId uint64, err error) {
```

Comment:
At brief glance, this method does not seem to support lock that was unbonding && undelegating, but only support lock that was locked but superfluid delegating. 
Can you confirm if this is the case or not?

Reply:
This is not the case. The following test case also tests this:

https://github.com/osmosis-labs/osmosis/blob/839ea2c08c129388e1ced8d68e91ed1cfa7c33b9/x/superfluid/keeper/migrate_test.go#L61-L66